ambry-rest/src/main/java/com.github.ambry.rest/NettySslHttp2Factory.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2018 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.rest;
+
+import com.github.ambry.commons.JdkSslFactory;
+import com.github.ambry.commons.SSLFactory;
+import com.github.ambry.config.SSLConfig;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.handler.codec.http2.Http2SecurityUtil;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.ApplicationProtocolNames;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SupportedCipherSuiteFilter;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.rest.NettySslFactory.*;
+
+
+/**
+ * An HTTP/2 specific implementation of {@link SSLFactory} that uses Netty's SSL libraries for HTTP2.
+ * This has the benefit of using OpenSSL instead of JDK's SSL implementation when the netty-tcnative library is loaded.
+ * OpenSSL shows significant performance enhancements over the JDK implementation.
+ */
+public class NettySslHttp2Factory implements SSLFactory {
+  private static final Logger logger = LoggerFactory.getLogger(NettySslFactory.class);
+  private final SslContext nettyServerSslContext;
+  private final SslContext nettyClientSslContext;
+  private final SSLContext jdkSslContext;
+  private final String endpointIdentification;
+
+  /**
+   * Instantiate a {@link NettySslFactory} from a config.
+   * @param sslConfig the {@link SSLConfig} to use.
+   * @throws GeneralSecurityException
+   * @throws IOException
+   */
+  public NettySslHttp2Factory(SSLConfig sslConfig) throws GeneralSecurityException, IOException {
+    nettyServerSslContext = getServerSslContext(sslConfig);
+    nettyClientSslContext = getClientSslContext(sslConfig);
+    // Netty's OpenSsl based implementation does not use the JDK SSLContext so we have to fall back to the JDK based
+    // factory to support this method.
+
+    jdkSslContext = sslConfig.sslHttp2SelfSign ? null : new JdkSslFactory(sslConfig).getSSLContext();
```

Comment:
I think there may not be a valid case for supporting the `getSSLContext()` method for HTTP2, since it was mostly present so that `BlockingChannelConnectionPool` could use blocking `SSLSocket`s instead of `SSLEngine`s, but there won't be that need with HTTP2. Instead, `getSSLContext()` could just throw `UnsupportedOperationException`.